### DISCUSSION TITLE
Buffered SD Card player

### DISCUSSION THREAD

**Discussion by madskjeldgaard** [pschatzmann/arduino-audio-tools#1726] [Ideas]
Hi all

I've run into some issues with playing .wav-files on both the Pico using Arduino Audio Tools as well as the Teensy using the Teensy Audio Lib. A big part of the problem in both instances seems to be the lack of reliability in the read speeds from an SD card and the high demand for data in a realtime audio context. The solution for Teensy was to create a Buffered SD Card WAV player class which works really well and has really nice performance.  something like this would be amazing in the arduino audio tools toolbox since it would allow to more easily stream high sample rate and uncompressed audio files from an SD card using SPI (and also make recording to SD card realible).

See it here on this PR: 

https://github.com/PaulStoffregen/Audio/pull/483

**Comment by madskjeldgaard** [comment]
The key classes for the Teensy version may be found here:

*Playing Wav files*

https://github.com/h4yn0nnym0u5e/Audio/blob/feature/buffered-SD-for-PR/play_wav_buffered.cpp

*Recording*

https://github.com/h4yn0nnym0u5e/Audio/blob/feature/buffered-SD-for-PR/record_wav_buffered.cpp

*The buffer*

https://github.com/h4yn0nnym0u5e/Audio/blob/feature/buffered-SD-for-PR/AudioBuffer.cpp

**Comment by pschatzmann** [comment]
In the Audiotools, usually the output classes have some buffer that you can increase.
I also recommend to maximize the SPI frequency or to use a 4 pin SDMMC to increase the SD thruput.
With this I could eaily play a wav file with 24 bit stereo audio at 48000 samples per second to I2S!

You can also easily implement a buffer of arbitrary size with following strategy:
- Create task which keeps a synchronized buffer (e.g. BufferRTOS) filled 
- Use separate task to consume from the buffer
 
Further info can be [found in the Wiki](https://github.com/pschatzmann/arduino-audio-tools/wiki/Multicore-Processing)
